// @ctrl/ngx-rightclick/esm5/context-submenu-trigger.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Directive, HostListener, Input, Output, EventEmitter, } from '@angular/core';
import { ContextMenuService, } from './context-menu.service';
var ContextSubmenuTriggerDirective = /** @class */ (function () {
    function ContextSubmenuTriggerDirective(contextMenuService) {
        var _this = this;
        this.contextMenuService = contextMenuService;
        this.hoverDelay = 500;
        this.openDelay = 200;
        this.menuAction = new EventEmitter();
        this.menuClose = new EventEmitter();
        this.visible = false;
        this.level = 1;
        // get current level
        setTimeout(function () { return (_this.level = _this.contextMenuService.getCurrentLevel()); });
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    ContextSubmenuTriggerDirective.prototype.handleSubMenuClick = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        event.preventDefault();
        event.stopPropagation();
        clearTimeout(this.opentimer);
        clearTimeout(this.closetimer);
        this.menu = this.contextMenuService.show($event, this.contextSubmenuTrigger, this.menuContext, this.menuClose, this.menuAction, true, this.level);
        this.visible = true;
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    ContextSubmenuTriggerDirective.prototype.handleSubMenuEnter = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        var _this = this;
        if (this.menu) {
            this.menu.isTriggerHovered.next(true);
        }
        clearTimeout(this.closetimer);
        this.opentimer = setTimeout(function () {
            _this.menu = _this.contextMenuService.show($event, _this.contextSubmenuTrigger, _this.menuContext, _this.menuClose, _this.menuAction, true, _this.level);
            _this.visible = true;
            _this.opentimer = null;
        }, this.openDelay);
    };
    /**
     * submenu hides after cursor has exited for a period of time
     */
    /**
     * submenu hides after cursor has exited for a period of time
     * @return {?}
     */
    ContextSubmenuTriggerDirective.prototype.handleSubMenuExit = /**
     * submenu hides after cursor has exited for a period of time
     * @return {?}
     */
    function () {
        var _this = this;
        clearTimeout(this.opentimer);
        if (this.menu) {
            this.menu.isTriggerHovered.next(false);
        }
        this.closetimer = setTimeout(function () {
            if (_this.menu) {
                _this.menu.isTriggerHovered.next(false);
                _this.contextMenuService.closeSubMenu(_this.menu.id);
                _this.menu = undefined;
            }
            _this.visible = false;
        }, this.hoverDelay);
    };
    /**
     * if overwritten make sure to clear timeouts
     */
    /**
     * if overwritten make sure to clear timeouts
     * @return {?}
     */
    ContextSubmenuTriggerDirective.prototype.ngOnDestroy = /**
     * if overwritten make sure to clear timeouts
     * @return {?}
     */
    function () {
        clearTimeout(this.opentimer);
        clearTimeout(this.closetimer);
    };
    ContextSubmenuTriggerDirective.decorators = [
        { type: Directive, args: [{ selector: '[contextSubmenuTrigger]' },] }
    ];
    /** @nocollapse */
    ContextSubmenuTriggerDirective.ctorParameters = function () { return [
        { type: ContextMenuService }
    ]; };
    ContextSubmenuTriggerDirective.propDecorators = {
        hoverDelay: [{ type: Input }],
        openDelay: [{ type: Input }],
        contextSubmenuTrigger: [{ type: Input }],
        menuContext: [{ type: Input }],
        menuAction: [{ type: Output }],
        menuClose: [{ type: Output }],
        handleSubMenuClick: [{ type: HostListener, args: ['click', ['$event', 'true'],] }],
        handleSubMenuEnter: [{ type: HostListener, args: ['mouseover', ['$event'],] }],
        handleSubMenuExit: [{ type: HostListener, args: ['mouseout', ['$event'],] }]
    };
    return ContextSubmenuTriggerDirective;
}());
export { ContextSubmenuTriggerDirective };
if (false) {
    /** @type {?} */
    ContextSubmenuTriggerDirective.prototype.hoverDelay;
    /** @type {?} */
    ContextSubmenuTriggerDirective.prototype.openDelay;
    /** @type {?} */
    ContextSubmenuTriggerDirective.prototype.contextSubmenuTrigger;
    /** @type {?} */
    ContextSubmenuTriggerDirective.prototype.menuContext;
    /** @type {?} */
    ContextSubmenuTriggerDirective.prototype.menuAction;
    /** @type {?} */
    ContextSubmenuTriggerDirective.prototype.menuClose;
    /** @type {?} */
    ContextSubmenuTriggerDirective.prototype.menu;
    /** @type {?} */
    ContextSubmenuTriggerDirective.prototype.opentimer;
    /** @type {?} */
    ContextSubmenuTriggerDirective.prototype.closetimer;
    /** @type {?} */
    ContextSubmenuTriggerDirective.prototype.visible;
    /** @type {?} */
    ContextSubmenuTriggerDirective.prototype.level;
    /** @type {?} */
    ContextSubmenuTriggerDirective.prototype.contextMenuService;
}
