// @ctrl/ngx-rightclick/esm5/context-menu-trigger.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Directive, HostListener, Input, Output, EventEmitter, } from '@angular/core';
import { ContextMenuService } from './context-menu.service';
var ContextMenuTriggerDirective = /** @class */ (function () {
    function ContextMenuTriggerDirective(contextMenuService) {
        this.contextMenuService = contextMenuService;
        this.holdToDisplay = 1000;
        this.menuAction = new EventEmitter();
        this.menuClose = new EventEmitter();
        this.visible = false;
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    ContextMenuTriggerDirective.prototype.handleMenu = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        $event.preventDefault();
        this.menu = this.contextMenuService.show($event, this.contextMenuTrigger, this.menuContext, this.menuClose, this.menuAction);
        this.visible = true;
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    ContextMenuTriggerDirective.prototype.handleMouseDown = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        var _this = this;
        if (this.holdToDisplay >= 0) {
            $event.stopPropagation();
            $event.clientY = $event.touches[0].clientY;
            $event.clientX = $event.touches[0].clientX;
            this.mouseDownTimeoutId = setTimeout(function () { return _this.handleMenu($event); }, this.holdToDisplay);
        }
    };
    /**
     * @return {?}
     */
    ContextMenuTriggerDirective.prototype.handleMouseUp = /**
     * @return {?}
     */
    function () {
        clearTimeout(this.mouseDownTimeoutId);
    };
    /**
     * @return {?}
     */
    ContextMenuTriggerDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.sub = this.menuClose.subscribe(function () { return _this.visible = false; });
    };
    /**
     * @return {?}
     */
    ContextMenuTriggerDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.sub.unsubscribe();
    };
    ContextMenuTriggerDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[contextMenuTrigger]',
                    exportAs: 'contextMenuTrigger',
                },] }
    ];
    /** @nocollapse */
    ContextMenuTriggerDirective.ctorParameters = function () { return [
        { type: ContextMenuService }
    ]; };
    ContextMenuTriggerDirective.propDecorators = {
        contextMenuTrigger: [{ type: Input }],
        menuContext: [{ type: Input }],
        holdToDisplay: [{ type: Input }],
        menuAction: [{ type: Output }],
        menuClose: [{ type: Output }],
        handleMenu: [{ type: HostListener, args: ['contextmenu', ['$event'],] }],
        handleMouseDown: [{ type: HostListener, args: ['touchstart', ['$event'],] }],
        handleMouseUp: [{ type: HostListener, args: ['touchend',] }]
    };
    return ContextMenuTriggerDirective;
}());
export { ContextMenuTriggerDirective };
if (false) {
    /** @type {?} */
    ContextMenuTriggerDirective.prototype.contextMenuTrigger;
    /** @type {?} */
    ContextMenuTriggerDirective.prototype.menuContext;
    /** @type {?} */
    ContextMenuTriggerDirective.prototype.holdToDisplay;
    /** @type {?} */
    ContextMenuTriggerDirective.prototype.menuAction;
    /** @type {?} */
    ContextMenuTriggerDirective.prototype.menuClose;
    /** @type {?} */
    ContextMenuTriggerDirective.prototype.menu;
    /** @type {?} */
    ContextMenuTriggerDirective.prototype.visible;
    /** @type {?} */
    ContextMenuTriggerDirective.prototype.mouseDownTimeoutId;
    /** @type {?} */
    ContextMenuTriggerDirective.prototype.sub;
    /** @type {?} */
    ContextMenuTriggerDirective.prototype.contextMenuService;
}
