// @ctrl/ngx-rightclick/esm5/menu.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, HostListener } from '@angular/core';
import { ContextMenuService } from './context-menu.service';
import { MenuPackage } from './context-menu-injector';
import { Subject } from 'rxjs';
var MenuComponent = /** @class */ (function () {
    function MenuComponent(menuPackage, contextMenuService) {
        this.menuPackage = menuPackage;
        this.contextMenuService = contextMenuService;
        /**
         * State of the dialog animation.
         */
        this._state = 'enter';
        this._animationDone = new Subject();
        /**
         * set lazy to False if you do not have animations
         */
        this.lazy = true;
    }
    /**
     * @return {?}
     */
    MenuComponent.prototype.handleMouseover = /**
     * @return {?}
     */
    function () {
        if (!this.menuPackage.menu.submenu) {
            return;
        }
        this.menuPackage.menu.isMenuHovered.next(true);
        clearTimeout(this.closetimer);
    };
    /**
     * @return {?}
     */
    MenuComponent.prototype.handleMouseleave = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.menuPackage.menu.submenu) {
            return;
        }
        this.menuPackage.menu.isMenuHovered.next(false);
        this.closetimer = setTimeout(function () {
            _this.contextMenuService.closeSubMenu(_this.menuPackage.menu.id);
        }, 500);
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    MenuComponent.prototype.handleWindowClick = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.contextMenuService.checkOutsideClick($event);
    };
    /** Callback that is invoked when the menu animation completes. */
    /**
     * Callback that is invoked when the menu animation completes.
     * @param {?} event
     * @return {?}
     */
    MenuComponent.prototype._onAnimationDone = /**
     * Callback that is invoked when the menu animation completes.
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this._animationDone.next(event);
    };
    MenuComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-menu',
                    template: ""
                }] }
    ];
    /** @nocollapse */
    MenuComponent.ctorParameters = function () { return [
        { type: MenuPackage },
        { type: ContextMenuService }
    ]; };
    MenuComponent.propDecorators = {
        handleMouseover: [{ type: HostListener, args: ['mouseover',] }],
        handleMouseleave: [{ type: HostListener, args: ['mouseleave',] }],
        handleWindowClick: [{ type: HostListener, args: ['document:click', ['$event'],] }]
    };
    return MenuComponent;
}());
export { MenuComponent };
if (false) {
    /**
     * State of the dialog animation.
     * @type {?}
     */
    MenuComponent.prototype._state;
    /** @type {?} */
    MenuComponent.prototype._animationDone;
    /**
     * set lazy to False if you do not have animations
     * @type {?}
     */
    MenuComponent.prototype.lazy;
    /** @type {?} */
    MenuComponent.prototype.closetimer;
    /** @type {?} */
    MenuComponent.prototype.menuPackage;
    /** @type {?} */
    MenuComponent.prototype.contextMenuService;
}
