// @ctrl/ngx-rightclick/esm2015/context-submenu-trigger.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Directive, HostListener, Input, Output, EventEmitter, } from '@angular/core';
import { ContextMenuService, } from './context-menu.service';
export class ContextSubmenuTriggerDirective {
    /**
     * @param {?} contextMenuService
     */
    constructor(contextMenuService) {
        this.contextMenuService = contextMenuService;
        this.hoverDelay = 500;
        this.openDelay = 200;
        this.menuAction = new EventEmitter();
        this.menuClose = new EventEmitter();
        this.visible = false;
        this.level = 1;
        // get current level
        setTimeout(() => (this.level = this.contextMenuService.getCurrentLevel()));
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    handleSubMenuClick($event) {
        event.preventDefault();
        event.stopPropagation();
        clearTimeout(this.opentimer);
        clearTimeout(this.closetimer);
        this.menu = this.contextMenuService.show($event, this.contextSubmenuTrigger, this.menuContext, this.menuClose, this.menuAction, true, this.level);
        this.visible = true;
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    handleSubMenuEnter($event) {
        if (this.menu) {
            this.menu.isTriggerHovered.next(true);
        }
        clearTimeout(this.closetimer);
        this.opentimer = setTimeout(() => {
            this.menu = this.contextMenuService.show($event, this.contextSubmenuTrigger, this.menuContext, this.menuClose, this.menuAction, true, this.level);
            this.visible = true;
            this.opentimer = null;
        }, this.openDelay);
    }
    /**
     * submenu hides after cursor has exited for a period of time
     * @return {?}
     */
    handleSubMenuExit() {
        clearTimeout(this.opentimer);
        if (this.menu) {
            this.menu.isTriggerHovered.next(false);
        }
        this.closetimer = setTimeout(() => {
            if (this.menu) {
                this.menu.isTriggerHovered.next(false);
                this.contextMenuService.closeSubMenu(this.menu.id);
                this.menu = undefined;
            }
            this.visible = false;
        }, this.hoverDelay);
    }
    /**
     * if overwritten make sure to clear timeouts
     * @return {?}
     */
    ngOnDestroy() {
        clearTimeout(this.opentimer);
        clearTimeout(this.closetimer);
    }
}
ContextSubmenuTriggerDirective.decorators = [
    { type: Directive, args: [{ selector: '[contextSubmenuTrigger]' },] }
];
/** @nocollapse */
ContextSubmenuTriggerDirective.ctorParameters = () => [
    { type: ContextMenuService }
];
ContextSubmenuTriggerDirective.propDecorators = {
    hoverDelay: [{ type: Input }],
    openDelay: [{ type: Input }],
    contextSubmenuTrigger: [{ type: Input }],
    menuContext: [{ type: Input }],
    menuAction: [{ type: Output }],
    menuClose: [{ type: Output }],
    handleSubMenuClick: [{ type: HostListener, args: ['click', ['$event', 'true'],] }],
    handleSubMenuEnter: [{ type: HostListener, args: ['mouseover', ['$event'],] }],
    handleSubMenuExit: [{ type: HostListener, args: ['mouseout', ['$event'],] }]
};
if (false) {
    /** @type {?} */
    ContextSubmenuTriggerDirective.prototype.hoverDelay;
    /** @type {?} */
    ContextSubmenuTriggerDirective.prototype.openDelay;
    /** @type {?} */
    ContextSubmenuTriggerDirective.prototype.contextSubmenuTrigger;
    /** @type {?} */
    ContextSubmenuTriggerDirective.prototype.menuContext;
    /** @type {?} */
    ContextSubmenuTriggerDirective.prototype.menuAction;
    /** @type {?} */
    ContextSubmenuTriggerDirective.prototype.menuClose;
    /** @type {?} */
    ContextSubmenuTriggerDirective.prototype.menu;
    /** @type {?} */
    ContextSubmenuTriggerDirective.prototype.opentimer;
    /** @type {?} */
    ContextSubmenuTriggerDirective.prototype.closetimer;
    /** @type {?} */
    ContextSubmenuTriggerDirective.prototype.visible;
    /** @type {?} */
    ContextSubmenuTriggerDirective.prototype.level;
    /** @type {?} */
    ContextSubmenuTriggerDirective.prototype.contextMenuService;
}
