// @ctrl/ngx-rightclick/esm2015/context-menu-trigger.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Directive, HostListener, Input, Output, EventEmitter, } from '@angular/core';
import { ContextMenuService } from './context-menu.service';
export class ContextMenuTriggerDirective {
    /**
     * @param {?} contextMenuService
     */
    constructor(contextMenuService) {
        this.contextMenuService = contextMenuService;
        this.holdToDisplay = 1000;
        this.menuAction = new EventEmitter();
        this.menuClose = new EventEmitter();
        this.visible = false;
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    handleMenu($event) {
        $event.preventDefault();
        this.menu = this.contextMenuService.show($event, this.contextMenuTrigger, this.menuContext, this.menuClose, this.menuAction);
        this.visible = true;
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    handleMouseDown($event) {
        if (this.holdToDisplay >= 0) {
            $event.stopPropagation();
            $event.clientY = $event.touches[0].clientY;
            $event.clientX = $event.touches[0].clientX;
            this.mouseDownTimeoutId = setTimeout(() => this.handleMenu($event), this.holdToDisplay);
        }
    }
    /**
     * @return {?}
     */
    handleMouseUp() {
        clearTimeout(this.mouseDownTimeoutId);
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.sub = this.menuClose.subscribe(() => this.visible = false);
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.sub.unsubscribe();
    }
}
ContextMenuTriggerDirective.decorators = [
    { type: Directive, args: [{
                selector: '[contextMenuTrigger]',
                exportAs: 'contextMenuTrigger',
            },] }
];
/** @nocollapse */
ContextMenuTriggerDirective.ctorParameters = () => [
    { type: ContextMenuService }
];
ContextMenuTriggerDirective.propDecorators = {
    contextMenuTrigger: [{ type: Input }],
    menuContext: [{ type: Input }],
    holdToDisplay: [{ type: Input }],
    menuAction: [{ type: Output }],
    menuClose: [{ type: Output }],
    handleMenu: [{ type: HostListener, args: ['contextmenu', ['$event'],] }],
    handleMouseDown: [{ type: HostListener, args: ['touchstart', ['$event'],] }],
    handleMouseUp: [{ type: HostListener, args: ['touchend',] }]
};
if (false) {
    /** @type {?} */
    ContextMenuTriggerDirective.prototype.contextMenuTrigger;
    /** @type {?} */
    ContextMenuTriggerDirective.prototype.menuContext;
    /** @type {?} */
    ContextMenuTriggerDirective.prototype.holdToDisplay;
    /** @type {?} */
    ContextMenuTriggerDirective.prototype.menuAction;
    /** @type {?} */
    ContextMenuTriggerDirective.prototype.menuClose;
    /** @type {?} */
    ContextMenuTriggerDirective.prototype.menu;
    /** @type {?} */
    ContextMenuTriggerDirective.prototype.visible;
    /** @type {?} */
    ContextMenuTriggerDirective.prototype.mouseDownTimeoutId;
    /** @type {?} */
    ContextMenuTriggerDirective.prototype.sub;
    /** @type {?} */
    ContextMenuTriggerDirective.prototype.contextMenuService;
}
