// @ctrl/ngx-rightclick/esm2015/menu.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, HostListener } from '@angular/core';
import { ContextMenuService } from './context-menu.service';
import { MenuPackage } from './context-menu-injector';
import { Subject } from 'rxjs';
export class MenuComponent {
    /**
     * @param {?} menuPackage
     * @param {?} contextMenuService
     */
    constructor(menuPackage, contextMenuService) {
        this.menuPackage = menuPackage;
        this.contextMenuService = contextMenuService;
        /**
         * State of the dialog animation.
         */
        this._state = 'enter';
        this._animationDone = new Subject();
        /**
         * set lazy to False if you do not have animations
         */
        this.lazy = true;
    }
    /**
     * @return {?}
     */
    handleMouseover() {
        if (!this.menuPackage.menu.submenu) {
            return;
        }
        this.menuPackage.menu.isMenuHovered.next(true);
        clearTimeout(this.closetimer);
    }
    /**
     * @return {?}
     */
    handleMouseleave() {
        if (!this.menuPackage.menu.submenu) {
            return;
        }
        this.menuPackage.menu.isMenuHovered.next(false);
        this.closetimer = setTimeout(() => {
            this.contextMenuService.closeSubMenu(this.menuPackage.menu.id);
        }, 500);
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    handleWindowClick($event) {
        this.contextMenuService.checkOutsideClick($event);
    }
    /**
     * Callback that is invoked when the menu animation completes.
     * @param {?} event
     * @return {?}
     */
    _onAnimationDone(event) {
        this._animationDone.next(event);
    }
}
MenuComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-menu',
                template: ``
            }] }
];
/** @nocollapse */
MenuComponent.ctorParameters = () => [
    { type: MenuPackage },
    { type: ContextMenuService }
];
MenuComponent.propDecorators = {
    handleMouseover: [{ type: HostListener, args: ['mouseover',] }],
    handleMouseleave: [{ type: HostListener, args: ['mouseleave',] }],
    handleWindowClick: [{ type: HostListener, args: ['document:click', ['$event'],] }]
};
if (false) {
    /**
     * State of the dialog animation.
     * @type {?}
     */
    MenuComponent.prototype._state;
    /** @type {?} */
    MenuComponent.prototype._animationDone;
    /**
     * set lazy to False if you do not have animations
     * @type {?}
     */
    MenuComponent.prototype.lazy;
    /** @type {?} */
    MenuComponent.prototype.closetimer;
    /** @type {?} */
    MenuComponent.prototype.menuPackage;
    /** @type {?} */
    MenuComponent.prototype.contextMenuService;
}
